Rework purity flag setting for constant variables accessed via member access.

rodiazet opened the pull request:
In this PR:
- `TypeChecker` rework for `MemberAccess` implementation
  -  Move `isPure` setting for `VariableDeclaration` of constant variable to the context of accessing it via `TypeType`, not via `this`.
  - Rework `TypeType` case for `ContractType` for to properly handle events, errors, structs and enums when accessing via contract name.
  - The same as above but in context of module.
  - Enable removed previously sanity check which did not work
- Rework `nativeMembers` implementation and `typeViaContractName` to return function type with call kind properly set for any context.
- Add and update tests for member access.  

~Depends on https://github.com/argotorg/solidity/pull/16448~
Depends on: https://github.com/argotorg/solidity/pull/16626

cameel commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3247,22 +3247,71 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 	else if (exprType->category() == Type::Category::Module)
 	{
-		annotation.isPure = *_memberAccess.expression().annotation().isPure;
+		// Very similar as for `ContractType`, but additionally we have to handle `Mod.C` case, where `C` in contract
```
```suggestion
		// Very similar as for `ContractType`, but additionally we have to handle `Mod.C` case, where `C` is contract
```

cameel commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3247,22 +3247,71 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+				typeTypeMember->actualType()->category() == Type::Category::Struct ||
+				typeTypeMember->actualType()->category() == Type::Category::Enum ||
+				typeTypeMember->actualType()->category() == Type::Category::Contract
+			)
```
I would add `UserDefinedValueType` as well.

And if you add that I think this covers pretty much every category that is possible inside a module via language syntax. So I'd simply make it always true for `TypeType` and just keep the condition as an assert to make sure we're correct about this assumption.

cameel replied:
Same thing for the contract case. `UserDefinedValueType` and definitions are allowed inside contracts. `Contract` is not, but if it was, I should be marked pure as well, so I'd include it.

This makes the code the same as for modules. And this sounds about right - I don't think this logic is in any way specific to modules. `TypeType` should generally always be constant. This is what we already do when type-checking `Identifier`: https://github.com/argotorg/solidity/blob/2673c4b6179c681f6601b607ef64852cfca5b52c/libsolidity/analysis/TypeChecker.cpp#L3701-L3702

cameel commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3247,22 +3247,71 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+				functionTypeMember->isPure() ||
+				functionTypeMember->kind() == FunctionType::Kind::Internal ||
+				functionTypeMember->kind() == FunctionType::Kind::Event
+			)
```
I think this is missing `Error`. Same for the contract case.

And, again, with that this covers all possible function types you can have as members on a module, so you can just assert and reduce the condition to `if (functionTypeMember)`.

cameel replied:
Well, technically you could try to tack an `External` function onto a module this way:

```solidity
import "m.sol" as M;

library L {
    function f(uint x) external {}
}

contract C {
    using L for *;

    function test() public {
        M.f; // Error: Member "f" not found or not visible after argument-dependent lookup in module "m.sol".
    }
}
```

`using for` attaches the function to all types, even those unnameable, so you can use this to attach functions even to things which cannot otherwise be the type of `x`, e.g. to literals. And we cannot filter out non-matching types simply by checking if the type is the same, because we have to take into account implicit conversions so there is a chance that you could attach your function to weird things this way, even if you won't be able to call it.

Not sure if we're doing more complex filtering here or if it's just hardcoded to only accept some specific types, but either way this fortunately does not compile, so we do not have to worry about external functions after all.

rodiazet replied:
`Error` and `Declaration` are cover by `isPure`.

cameel replied:
Oh. Maybe we should add events there as well? Both errors and events by themselves should be pure (and even theoretically usable to initialize constants, though there's nothing you can do with them currently). If we introduce proper error/event objects, those should not require relaxing purity unless you actually send them to the outside in some way. It's only specifically `emit` and `revert` that should have higher requirements.

cameel commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3247,22 +3247,71 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+					functionTypeMember->isPure() ||
+					functionTypeMember->kind() == FunctionType::Kind::Internal ||
+					functionTypeMember->kind() == FunctionType::Kind::Event
+				)
```
I think this is missing `Error`.

cameel commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3247,22 +3247,71 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+			// See `ContractType::nativeMembers` for details.
+			solAssert(annotation.referencedDeclaration);
 			annotation.isLValue = annotation.referencedDeclaration->isLValue();
+			// Expressions like `C.foo`, `C.Ev` are pure and they must generate `Statement has no effect.` warning.
```
What about function pointers?

```solidity
contract C {
    function() internal returns (uint) fi;
    function() internal returns (uint) gi;
    function() external returns (uint) fe;
    function() external returns (uint) ge;

    function test() public {
        C.fi = C.gi;
        C.fe = C.ge;
    }
}
```

And generally it looks like you can refer to member variables via contract name, which means that you can have non-pure things as contract type members:
```solidity
contract C {
    uint x;
    uint y;

    function test() public {
        C.x = C.y;
    }
}
```

rodiazet replied:
The comment was a little unclear. I was thinking of `C.foo;`, `C.Ev;` statements. The same should apply to `C.fi;`, `C.gi;` ,`C.fe;`, `C.ge;`, `C.x;` and `C.y;`, the warning should be generated. Problem is that it's generated based on `isPure` flag, but this flag is also used to mark that an expression can be used to initialize a constant (compile-time) variable. 
If we set that all the expression have `isPure == true`, it makes 
```solidity
contract C {
    uint y;
    uint constant x = C.y;
}
```
compiles. Which is obviously wrong.

rodiazet replied:
I have decided  to leave internal call kind function as not pure. So `C.foo` is never pure. In the https://github.com/argotorg/solidity/pull/8263 the pure flag was set to true for external call kind functions accessed via contract name. Should we revert this or it's ok? @cameel 
For the function pointers we mark them pure only if the pointer variable is constant. (Similar to other variables.) For now it never happens because there is no way to initialize constant function pointer (https://github.com/argotorg/solidity/issues/16339).

cameel replied:
> I have decided to leave internal call kind function as not pure.

I'm fine with not touching them in this PR, but I'd also be fine with giving them `isPure = true` here. That's the end state I'd like to get eventually anyway (#16422).

I had doubts about that when I reported #16339, but that was because I was looking at `constant` as something that should be become comptime. Now I think we should keep it runtime constant and having inline functions be runtime constant as well would be perfectly in line with that.

> In the #8263 the pure flag was set to true for external call kind functions accessed via contract name. Should we revert this or it's ok?

Seems fine to me. They're a weird construct, since they're not associated with an address, but I can't see any reason why we would not treat them as runtime constant.

cameel commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3247,22 +3247,71 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+				varDeclMember->isConstant()
 			)
-				annotation.isPure = *_memberAccess.expression().annotation().isPure;
+				annotation.isPure = true;
```
Currently you're checking `isConstant()` only on contract types and modules, assuming that the only other possible case is access via a contract variable (which means it must be a getter function). This is probably true, but we should guard it with an assert which will fail if we missed any other case.

Or it might be simpler to just keep the check as it was and only specifically exclude the contract variable case.

rodiazet replied:
I prefer to add separated asset.

rodiazet replied:
BTW this can be also a variable declaration in a struct, accessed via a member access.

cameel commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3247,22 +3247,71 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 	}
 	else if (exprType->category() == Type::Category::Module)
 	{
-		annotation.isPure = *_memberAccess.expression().annotation().isPure;
```
You're only handling `FunctionType`, `TypeType` and things associated with a variable declaration. I think originally this covered more cases:
- `Category::Module` (i.e. module accessed through another module, `M1.M2`). Original code would make it pure because the member expression is an `Identifier` and those are pure when they refer to modules: https://github.com/argotorg/solidity/blob/2673c4b6179c681f6601b607ef64852cfca5b52c/libsolidity/analysis/TypeChecker.cpp#L3703-L3704
    After your changes I think we'll reach the `!annotation.isPure.set()` at the end of the function instead and set it to false.
- `MagicVariableDeclaration`. I think this one cannot occur on a module, but I'd assert against that.
- There are probably no other possibilities (AFAIR modules can contain functions, constants, events, errors and type definitions) but I'd assert against that too.

cameel commented on libsolidity/ast/AST.cpp:
```diff
@@ -491,17 +491,31 @@ Type const* FunctionDefinition::type() const
 	if (libraryFunction())
 	{
 		if (isPublic())
+			// Lib.foo is public or external, an external call is used.
```
```suggestion
			// When Lib.foo is public or external, an external call is used.
```

cameel commented on libsolidity/ast/AST.h:
```diff
@@ -299,7 +299,7 @@ class Declaration: public ASTNode, public Scopable
 	/// @returns the type for members of the containing contract type that refer to this declaration.
 	/// This can only be called once types of variable declarations have already been resolved.
-	virtual Type const* typeViaContractName() const { return type(); }
+	virtual Type const* typeViaContractName(bool) const { return type(); }
```
The new argument should be documented.

Also, what exactly does it mean? `inDerivingScope` sounds a little ambiguous, maybe the name should be adjusted? Is it saying whether we're still within the scope that contains the definition or something? EDIT: After reading further I see that's a preexisting name and literally refers to the derived contract. How about calling the parameter `inherited`?

I would also consider making it an enum. These boolean args we already have are not very readable at the point of call and I usually suggest adding a comment that mentions the parameter name when calling the function anyway. An enum would make that unnecessary.

rodiazet replied:
Reworked. I separated additional function `typeViaForeignContractName` and removed the flag.

cameel commented on libsolidity/ast/AST.cpp:
```diff
@@ -491,17 +491,31 @@ Type const* FunctionDefinition::type() const
+	solAssert(
+		libraryFunction() || visibility() != Visibility::Private,
+		"Private function members are not available via contract type name."
+	);
```
Maybe also assert that if `inDerivedScope`, the function cannot be a library function? That combination would not make sense.

And `isVisibleInDerivedContracts()` must also be true for that function. On the other hand when not in derived scope, `isVisibleInContract()` should still be true.

cameel replied:
Also, when `libraryFunction()`, assert that `isVisibleAsLibraryMember()` is true. This will not let you call this with private library functions, but those are not callable as `Lib.foo()`, only `foo()` anyway.

And that `isImplemented()` is true for library functions.

rodiazet replied:
There are cases when accessing private library members is allowed. I.e. in `using` statement.

rodiazet replied:
We also cannot `isImplemented` at this stage as it's checked later in [`TypeChecker::visit(FunctionDefinition const& _function)`](https://github.com/argotorg/solidity/blob/v0.8.33/libsolidity/analysis/TypeChecker.cpp#L456). Test triggering this error `syntaxTests/nameAndTypeResolution/229_call_to_library_function`

cameel commented on libsolidity/ast/Types.cpp:
```diff
@@ -3976,21 +3978,27 @@ MemberList::MemberMap TypeType::nativeMembers(ASTNode const* _currentScope) cons
+				{
+					// In case of regular contract being accessed from by external contract (not in deriving scope),
+					// add only externally visible members.
+					if (declaration->isVisibleViaContractTypeAccess())
```
We should rename this function to something like `isVisibleViaContractInstance()`. I thought it was referring to access like `C.foo()` until I looked at the definition.

rodiazet replied:
Done

cameel commented on libsolidity/ast/Types.cpp:
```diff
@@ -3976,21 +3978,27 @@ MemberList::MemberMap TypeType::nativeMembers(ASTNode const* _currentScope) cons
+					// In case of regular contract being accessed from by external contract (not in deriving scope),
+					// add only externally visible members.
+					if (declaration->isVisibleViaContractTypeAccess())
+						members.emplace_back(declaration, declaration->typeViaContractName(inDerivingScope));
```
All three branches here now call `typeViaContractName()`, so you could simplify it all into a single condition.

rodiazet replied:
Not any more. :)

cameel commented on libsolidity/ast/Types.cpp:
```diff
@@ -3976,21 +3978,27 @@ MemberList::MemberMap TypeType::nativeMembers(ASTNode const* _currentScope) cons
+					// In case of regular contract (not library) and member is in the same deriving scope, add all
+					// members which are not private. Private members cannot be accessed via contract type name
+					// i.e C.fooPrivate.
+					if (declaration->visibility() > Visibility::Private)
```
I think this will add external functions while previously `isVisibleInDerivedContracts()` would filter them out.

rodiazet replied:
Yes. And the Declaration kind was set below. Now it's handled by the `typeViaContractName(ContractNameAccessKind _accessKind)`

cameel commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,160 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+				exprType->category() == Type::Category::Struct ||
+				exprType->category() == Type::Category::Contract,
+				"Variable member is only available for module, contract or struct");
+		}
```
Braces are unnecessary here (and IMO just make it less readable).

The message could also be more precise. On the other hand it's superfluous to have it in both asserts as well as in a comment. I'd leave it just in the comment.
```suggestion
		else
			solAssert(
				exprType->category() == Type::Category::Module ||
				exprType->category() == Type::Category::Struct ||
				exprType->category() == Type::Category::Contract,
				"Only module, struct and contract instances as well as contract types can have accessible variables."
			);
```

Overall, assert messages are optional and I'd omit them for brevity in cases where the condition is obvious or already explained. And if it's not I'd usually rather try to make it clearer by introducing some intermediate variables with good names (though TBH naming in this part of code is pretty bad; e.g. `exprType` tells you nothing out of context, you have to find the declaration to figure out which particular expression it represents).

rodiazet replied:
Changed.

cameel commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,160 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+			);
+		}
+		else if (auto const* varDecl = dynamic_cast<VariableDeclaration const*>(annotation.referencedDeclaration))
+			annotation.isPure = *_memberAccess.expression().annotation().isPure && varDecl->isConstant();
```
This looks wrong. Did you mean `||`? The way it is now, only `constants` accessed via a constant contract instance will be marked constant.

This means this will not compile, because `c.x` won't be considered constant, even though the thing you're accessing is really a getter, not the constant itself:

```solidity
contract C {
    uint public x;
}

contract D {
    C constant c;
    function() external returns (uint) constant fPtr = c.x;
}
```

cameel replied:
Also, currently the only things I can think of that would fall into this branch are getters. I'd add an assert that `annotation.type` is `FunctionType` to verify that.

Actually, we'd now have it in both branches so we can pull it out of the `if`. It sounds about right. The only things you can access via a contract instance are its external functions (explicitly declared or getters) or library functions attached with `using for`.

cameel commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,160 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+	{
 		annotation.isLValue = !structType->dataStoredIn(DataLocation::CallData);
+
+		if (auto const* varDecl = dynamic_cast<VariableDeclaration const*>(annotation.referencedDeclaration))
```
This being called `varDecl` (sometimes even `vDecl`) is confusing me to no end every time I read this code, especially now that you're using it a lot more than it was before. With these names being this vague it's hard to keep track of what refers to the member, what to the underlying expression and what to the member access expression as a whole.

It's a bit awkward, because we don't have clear terms to distinguish these things, but maybe `declarationOfAccessedVariable` would be good enough for this one. And not just here but in all cases inside in this function.

```suggestion
		if (auto const* memberAccessVariableDeclaration = dynamic_cast<VariableDeclaration const*>(annotation.referencedDeclaration))
```

Similarly, I'd rename `funcType`/`functionTypeMember` to `accessedFunctionType`.

And `exprType` would have been much better off as `underlyingExpressionType` or something, though I'd probably not change it here because it's used a lot in the part of the function you're not touching. Doing that would introduce too much noise to the diff here and make reviewing it harder. So just FYI.

cameel replied:
Actually, can this condition even fail? Is it possible to have a struct-type expression without an associated struct definition? Shouldn't this be an assert?

rodiazet replied:
It's not about the struct, but the member of a structure. It can be also a function, which btw it not cover here. The function can be attached with `using` statement and a library. Have to add this too.

cameel commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,160 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+		if (auto const* varDecl = dynamic_cast<VariableDeclaration const*>(annotation.referencedDeclaration))
+		{
+			annotation.isPure = varDecl->isConstant();
+			solAssert(!varDecl->isConstant(), "Struct member variables cannot be declared as constant.");
```
I'd use `solUnimplemented` here to make it clear that it's not something that should not happen, but just something that we have not enabled yet.

```suggestion
			solUnimplementedAssert(!varDecl->isConstant(), "Constant structs are not yet implemented.");
```

cameel replied:
Also, why set `isPure` based on `isConstant()` in the first place? `_memberAccess.expression().annotation().isPure` seems more appropriate and future-proof to me. If we happen to have a constant struct instance, its members should also be constant. It would not require making assumptions about why it's constant (whether it's because `isConstant()` or due to some other mechanism we might introduce in the future).

cameel commented on libsolidity/analysis/TypeChecker.cpp:
We already have `else if (TypeType ...)` above. Perhaps we should merge them so that it's easier to follow?

In the same vein, the `FunctionType` and `MagicType` cases below look like they should really be just cases of that big `if` above. In fact, it should probably be just a `switch` on `annotation.type->category()`.

cameel replied:
Actually, what would you say to do a quick refactor of this in a separate PR to clean it up first and then rebase this PR on it? One problem I have with these conditions being restructured is that it will be hard to accurately tell which things visible to the user actually changed. We'll need that for changelog and docs. So it would be best if we could leave only actual behavioral changes here.

I'd also put the newly added tests in it. Then here we'll be able to easily see how the PR affects them.

rodiazet replied:
Good point. I was also thinking of it, because this PR is already getting too big.

rodiazet replied:
https://github.com/argotorg/solidity/pull/16448

cameel commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,160 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+		else if (annotation.type->category() == ModuleType::Category::Module)
+			annotation.isPure = true;
+		else if (auto const* varDecl = dynamic_cast<VariableDeclaration const*>(annotation.referencedDeclaration))
+			annotation.isPure = varDecl->isConstant();
```
Here I'd also initialize it from `isPure` of the referenced declaration. We could keep `isConstant()` as a sanity check, though.

rodiazet replied:
`VariableDeclaration` does not have `isPure` field, also in its annotation. Previously it was taken from the parent expression, not from the member, which does not make sense in case when the parent is a module.

cameel reviewed (commented):
Here's another batch of comments.

Though my general feeling is that it would be good to extract some things into a refactor PR to make it easier to see what's actually changing and what's existing behavior (see my comment below).

cameel commented on libsolidity/analysis/TypeChecker.cpp:
The PR is missing a changelog entry. With how much it's changing, we'll probably need more than one.

matheusaaguiar commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,199 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+				reinterpret_cast<TypeType const*>(exprType)->actualType()->category() == Type::Category::Contract,
+				""
+			);
+		} else
```
```suggestion
		}
		else
```
Also we can remove the braces of the `if`, remove the empty string of the assert and keep it in a single line.

matheusaaguiar commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,199 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+				accessedVariableDeclaration->isConstant();
+		}
+		else
+			solAssert(false,"Impossible declaration type for contact member.");
```
```suggestion
			solAssert(false,"Invalid declaration type for contract member.");
```
Maybe invalid is better than impossible. Anyway we should never see this message (hopefully :)) so it doesn't matter much.

matheusaaguiar commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,199 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 		annotation.isLValue = !structType->dataStoredIn(DataLocation::CallData);
+
+		if (auto const* accessedVariableDeclaration =
+			dynamic_cast<VariableDeclaration const*>(annotation.referencedDeclaration))
```
```suggestion
			dynamic_cast<VariableDeclaration const*>(annotation.referencedDeclaration)
	)
```

matheusaaguiar commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,199 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+				"Impossible function call kind for struct type member."
+			);
+
+			// When struct is constant its members are also constant.
```
```suggestion

```
I think this is evident in the code (members receive the same "pureness" as the struct).

matheusaaguiar commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,199 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+		else
+			solAssert(
+				false,
+				"Struct must have all members defined and they must be variables declarations or functions definitions"
```
```suggestion
				"Invalid struct member access"
```
Could either remove or simplify the message, no need to explain too much, users are not supposed to see this.

matheusaaguiar commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,199 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+					dynamic_cast<VariableDeclaration const*>(annotation.referencedDeclaration) ||
+					dynamic_cast<FunctionDefinition const*>(annotation.referencedDeclaration),
+					"Impossible declaration type for internal function call kind"
+				);
```
```suggestion
				else
					solAssert(false, "Impossible declaration type for internal function call kind");
```

matheusaaguiar commented on libsolidity/ast/AST.cpp:
```diff
@@ -491,17 +491,47 @@ Type const* FunctionDefinition::type() const
+		}
+		case ContractNameAccessKind::Foreign:
+		{
+			solAssert(!libraryFunction(), "Non-library members cannot be accessed via library name.");
```
```suggestion
			solAssert(!libraryFunction(), "Library members can only be accessed via library name.");
```
I guess this should be the message?

matheusaaguiar commented on libsolidity/ast/Types.cpp:
```diff
@@ -3976,21 +3980,33 @@ MemberList::MemberMap TypeType::nativeMembers(ASTNode const* _currentScope) cons
-						members.emplace_back(declaration, declaration->typeViaContractName());
-					else
-						members.emplace_back(declaration, declaration->type());
+					// In case the contract is library, add only visible library members. visibility >= Internal.
```
I think that this and the next comments are not necessary.

matheusaaguiar reviewed (commented):
Just some nitpicks and suggestions. I will give another pass tomorrow after I check the visitor refactor PR.

clonker commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,199 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+		if (exprType->category() == Type::Category::TypeType)
+		{
+			solAssert(
+				reinterpret_cast<TypeType const*>(exprType)->actualType()->category() == Type::Category::Contract,
```
Uhhh. This is in general actually UB. Check this little demonstration :)

```cpp
#include <iostream>

struct Base1 {
    virtual ~Base1() = default;
    int x = 1;
};

struct Base2 {
    virtual ~Base2() = default;
    int y = 2;
};

struct Derived : Base1, Base2 {
    int z = 3;
};

int main() {
    Derived d;

    Base2* b2 = &d;

    Derived* bad = reinterpret_cast<Derived*>(b2);
    Derived* good = dynamic_cast<Derived*>(b2);

    std::cout << "bad->z  = " << bad->z  << "\n";
    std::cout << "good->z = " << good->z << "\n";
}
```

With single inheritance and without virtual base class it is mostly fine but there are no guarantees in the standard. This should really be a `dynamic_cast`.

rodiazet replied:
Ok, but we know the exact type of the pointer, because we checked the category. To cover the case when some potentially adds one more base class I would use `static_cast` here, which properly calculates objects offsets. WDYT?

clonker replied:
Since this is a sanity check I'd err on the side of caution and use `dynamic_cast`. The `static_cast` is only fine if we have a guaranteed `1:1` correspondence between category and actual type. The performance impact should be negligible in any case to take `dynamic_cast` here, don't you think?

rodiazet replied:
I was thinking about performance exactly here, but if it's negligible I will change it to `dynamic_cast`. What do you think about the rest of the cases where I used `reinterpret_cast`?

clonker replied:
I mean all the cases in which you used `reinterpret_cast` is UB. IMO this cannot go into the codebase. As for performance: benchmark it :) but to me it seems like premature optimization. Happy to be proven wrong though!

rodiazet replied:
I was thinking of changing the rest to 'static_cast'

rodiazet replied:
If we cannot assume that we have 1:1 mapping to category we should also check if the result of the cast is not null.

clonker replied:
True actually. Well in that case, `static_cast` away. The `reinterpret_cast` has to go though :P

clonker commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,199 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+		if (dynamic_cast<FunctionDefinition const*>(annotation.referencedDeclaration))
+		{
+			annotation.isPure = *_memberAccess.expression().annotation().isPure;
+			auto const* accessedMemberFunctionType = reinterpret_cast<FunctionType const*>(annotation.type);
```
UB, see above

rodiazet replied:
fixed

clonker commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,199 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+		else if (dynamic_cast<FunctionDefinition const*>(annotation.referencedDeclaration))
+		{
+			solAssert(annotation.type->category() == Type::Category::Function, "");
+			auto const* accessedMemberFunctionType = reinterpret_cast<FunctionType const*>(annotation.type);
```
UB, see above

rodiazet replied:
fixed

clonker commented on libsolidity/ast/AST.h:
```diff
@@ -275,7 +275,7 @@ class Declaration: public ASTNode, public Scopable
 	virtual bool isVisibleInDerivedContracts() const { return isVisibleInContract() && visibility() >= Visibility::Internal; }
 	bool isVisibleAsLibraryMember() const { return visibility() >= Visibility::Internal; }
-	virtual bool isVisibleViaContractTypeAccess() const { return false; }
+	virtual bool isVisibleViaContractInstantce() const { return false; }
```
```suggestion
	virtual bool isVisibleViaContractInstantee() const { return false; }
```

clonker commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3409,6 +3567,16 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
 	if (!annotation.isPure.set())
 		annotation.isPure = false;
 
+	// Sanity check. For function type pure flag should be equal to
```
should be equal to...?

clonker commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3239,33 +3239,199 @@ bool TypeChecker::visit(MemberAccess const& _memberAccess)
+			}
+			// In case `Base.value` or `Lib.value` and when `value` is constant, the expression is pure.
+			else if (auto const* varDecl = dynamic_cast<VariableDeclaration const*>(annotation.referencedDeclaration))
+				annotation.isPure = varDecl->isConstant();
```
is it intentional that there is no catch-all handling here? to future-proof, maybe we should add an assert that it is only one of the three cases that are handled?

rodiazet replied:
I will test it with the assert but I think I did it already and it was failing in some cases. Don't remember why now, but I will add proper comment if that's the case.

rodiazet replied:
It passes. Will make an analysis if it really cannot happen.

clonker replied:
I'd still suggest an 

```cpp
else
    yulAssert(false, "impossible state");
```

then :) with perhaps something more descriptive than `impossible state`

rodiazet replied:
Yes. I'm going to add this, but want to make sure that we won't brake anything as I did with other ICE in the past. :)

clonker reviewed (commented):
Some initial comments, mostly typos and stuff. Is the PR this PR depends on the refactoring one that cameel suggested? If so, could you update the base of this PR to point to that one so the diff becomes a bit more manageable? :)

rodiazet committed:
Rename `isVisibleViaContractTypeAccess` -> `isVisibleViaContractInstance`

rodiazet committed:
Rework `typeViaContractName` and `nativeMembers` fragment.

rodiazet committed:
Implement struct instance member case.

Struct instance member can be only variable or a library function attached to the struct. For now the only possible way to have constant/pure struct instance is by inline declaring i.e. `S(1).v`

rodiazet committed:
Member access via a contract type name.

- Foreign function declaration access via contract name are pure.
- Functions declaration from derived scope are pure. Can be used to initialise constant function pointer.
- Internal library function declaration are pure. Can be used to initialise constant function pointer.
- struct, enum, UDVT declarations are pure.
- Constant variable declaration in contract and library are pure.
- event declaration is pure

rodiazet committed:
Member access via a contract type name.

- Foreign function declaration access via contract name are pure.
- Functions declaration from derived scope are pure. Can be used to initialise constant function pointer.
- Internal library function declaration are pure. Can be used to initialise constant function pointer.
- struct, enum, UDVT declarations are pure.
- Constant variable declaration in contract and library are pure.
- event declaration is pure

rodiazet committed:
Member access via module type.

rodiazet committed:
Member access via module type.

rodiazet committed:
Member access via contract instance.

rodiazet committed:
Member access via contract instance.

rodiazet committed:
Add accessible variables sanity check.

rodiazet committed:
Add accessible variables sanity check.

rodiazet committed:
Add error arguments purity tests.

rodiazet committed:
Add error arguments purity tests.

rodiazet committed:
Add constant access getter test which reproduces the bug.

rodiazet committed:
Add constant access getter test which reproduces the bug.

rodiazet committed:
Add purity tests of different types and usage.

rodiazet committed:
Add purity tests of different types and usage.

stackenbotten3000 commented on libsolidity/analysis/TypeChecker.cpp:
```diff
@@ -3092,307 +3090,459 @@ void TypeChecker::endVisit(NewExpression const& _newExpression)
 
-	annotation.requiredLookup = requiredLookup;
+	accessedMemberAnnotation.requiredLookup = requiredLookup;
+	
```
Coding style error

rodiazet committed:
Fix `constantValueFunction` to properly support function returning more than one value.

rodiazet committed:
Fix `constantValueFunction` to properly support function returning more than one value.

rodiazet committed:
Add semantic tests.

rodiazet committed:
Add semantic tests.
